### ISSUE TITLE
pin-project does not interoperate well with #[cfg()]

### ISSUE THREAD

**Issue by jonhoo** [taiki-e/pin-project#68]
Take this struct for instance:
```rust
#[pin_project]
#[derive(Debug)]
pub struct Socket {
    #[cfg_attr(unix, pin)]
    #[cfg(unix)]
    inner: tokio::net::unix::UnixStream,
    #[cfg_attr(windows, pin)]
    #[cfg(windows)]
    inner: tokio_named_pipes::NamedPipe,
}
```
I get the error
```rust
error[E0124]: field `inner` is already declared
  --> src/io/socket.rs:29:5
   |
26 |     inner: tokio::net::unix::UnixStream,
   |     ----------------------------------- `inner` first declared here
...
29 |     inner: tokio_named_pipes::NamedPipe,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field already declared
```
Even though, as far as I can tell, this _should_ work. My guess here is that `pin-project` is run before the `#[cfg()]`s are evaluated, and ends up not propagating them, but I'm not sure?

**Comment by jonhoo** [comment]
For now, I'm working around this by placing the `#[cfg()]` on the whole struct, but that won't easily work for structs with more fields that are shared.

**Comment by taiki-e** [comment]
@jonhoo Thanks for the reporting!

I will investigate.

**Comment by taiki-e** [comment]
This is because I missed that the timing when the `cfg` attributes are applied differs between proc-macro-derive and proc-macro-attribute (https://github.com/rust-lang/rust/pull/44528), but, I'm not sure if this issue can be fixed immediately.

1. pin-project collects the type of field annotated with `#[pin]` attributes from the structure field and generates an `Unpin` implementation.
2. Since `cfg` is not applied at that time, pin-project collects all `#[pin]` fields and adds them to where-clause of the `Unpin` implementation.
For example, in the example you wrote, the generated Unpin implementation would be:
```rust
impl Unpin for Socket 
where 
    tokio::net::unix::UnixStream: Unpin,
    tokio_named_pipes::NamedPipe: Unpin,
{}
```
To fix this, we probably need to delegate the `Unpin` implementation to proc-macro-derive. 
(Other issues are just extracting cfg from the field and applying it, so it shouldn't be too difficult to fix: abf3700d2dbf7d8452629562d4b3b487c72b4eea)

So, for now, it is preferable to create a type alias / new type or create structs for each `cfg`.

```rust
#[cfg(unix)]
struct Inner(tokio::net::unix::UnixStream); 
// or `type Inner = tokio::net::unix::UnixStream;`

#[cfg(windows)]
struct Inner(tokio_named_pipes::NamedPipe); 
// or `type Inner = tokio_named_pipes::NamedPipe;`

#[pin_project]
#[derive(Debug)]
pub struct Socket {
    #[pin]
    inner: Inner,
}
```

**Comment by taiki-e** [comment]
However, I would like to fix this by the release of 0.4 (the other remaining blocker at this time is #26).

**Comment by taiki-e** [comment]
Filed #77

**Comment by taiki-e** [comment]
Published 0.4.0-alpha.10 which fixes this issue.

By the way, if you use `cfg` on fields, it works for other attributes used on that field, so you can use `#[pin]` directly instead of `#[cfg_attr(..., pin)]`.

```rust
#[pin_project]
#[derive(Debug)]
pub struct Socket {
    #[pin]
    #[cfg(unix)]
    inner: tokio::net::unix::UnixStream,
    #[pin]
    #[cfg(windows)]
    inner: tokio_named_pipes::NamedPipe,
}
```
